File: singleton-design-pattern/payment-gateway.go
package main

import (
	"fmt"
	"sync"
)

type PaymentGatewayManager struct{}

var instance *PaymentGatewayManager
var mtx sync.Mutex

func GetInstance() *PaymentGatewayManager {
	if instance == nil { // First check (no lock)
		mtx.Lock() // Acquire lock
		defer mtx.Unlock()

		if instance == nil { // Double-check inside lock
			fmt.Println("Payment Gateway Manager initialized.")
			instance = &PaymentGatewayManager{}
		}
	}
	return instance
}

func (pg *PaymentGatewayManager) ProcessPayment(amount float64) {
	fmt.Printf("Processing payment of $%.2f through the payment gateway.\n", amount)
}

func main() {
	pg1 := GetInstance()
	pg1.ProcessPayment(100.0)

	pg2 := GetInstance()

	if pg1 == pg2 {
		fmt.Println("Both instances are the same. Singleton pattern is working.")
	} else {
		fmt.Println("Singleton pattern is not working correctly.")
	}
}

File: singleton-design-pattern/payment-gateway.js
class PaymentGatewayManager {
    constructor() {
      if (PaymentGatewayManager.instance) {
        return PaymentGatewayManager.instance;
      }
      console.log("Payment Gateway Manager initialized.");
      PaymentGatewayManager.instance = this;
    }
  
    processPayment(amount) {
      console.log(`Processing payment of $${amount} through the payment gateway.`);
    }
  
    static getInstance() {
      if (!PaymentGatewayManager.instance) {
        PaymentGatewayManager.instance = new PaymentGatewayManager();
      }
      return PaymentGatewayManager.instance;
    }
  }
  
  // Main simulation
  const pg1 = PaymentGatewayManager.getInstance();
  pg1.processPayment(100.0);
  
  const pg2 = PaymentGatewayManager.getInstance();
  
  if (pg1 === pg2) {
    console.log("Both instances are the same. Singleton pattern is working.");
  } else {
    console.log("Singleton pattern is not working correctly.");
  }

  
  // we dont need DCL here - because JS is single threaded